src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from "react";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import { PASSWORD_KEY } from "../constants";
+
+type AppModeTypes = "normal" | "kill";
+type AuthStatusTypes = "pending" | "authenticated" | "notRequired";
+
+type AuthState = {
+  passcode: string | null;
+  appMode: AppModeTypes;
+  killModeEnabled: boolean;
+  authStatus: AuthStatusTypes;
+};
+
+type KillModeActions =
+  | { type: "setPasscode"; newPasscode: string | null }
+  | { type: "toggleAppMode"; newAppMode?: AppModeTypes }
+  | { type: "toggleKillModeEnabled"; newKillModeValue?: boolean }
```

Comment:
after reading the implementations for these toggle actions, it might be clearer to separate them each into two actions, one for toggling and the other for setting a specific value. this way you avoid more complex logic checks based on the payload and state when the intention is purely to toggle. e.g. for toggling app mode:

```ts
// I like the idea of namespacing actions a bit but totally optional
| { type: 'appMode:toggle' }
| { type: 'appMode:set', newAppMode: AppModeTypes } // note that newAppMode is no longer optional here
```

Reply:
This feels like we are adding alot of extra noise....I think I will just use if statments to clear up the double ternary if that is the problem.

Reply:
the problem for me is that the intent of the action doesn't match the logic that happens within it. from a strict sense of "toggling", there shouldn't be a custom value to be provided like the current implementation has it as of right now. setting a value is a very different action than toggling to me. the optionality of a specific value makes the logic more unclear as a result.

having more actions if the actions are clearly defined and less logic to implement feels like a okay tradeoff

Reply:
I see what you mean. I am worried about over complicating things to have 2 state setters, for the same state, especially when there is only 2 possible states. If we changed the name from `toggle` to `setOrToggle` would that be a good compromise? Or do you think have multiple state setters is the better solution

Reply:
still think having multiple setters is still the clearer solution in this case. there isn't much overhead when it comes to adding actions compared to writing logic to handle complicated and potentially unclear actions. this becomes more apparent when there are more than the 2 states we're starting off with (can imagine this going to 3+ states at some point)